### ISSUE TITLE
On calculation of the score for the Gaussian likelihood

### ISSUE THREAD

**Issue by mengxiangming** [utcsilab/csgm-mri-langevin#4]
Hi, I am writing to consult you on the computation of likelihood score $\nabla_{x_{t}}\log\mu(y\mid x_{t})=\frac{A^{H}(y-Ax_{t})}{\gamma_{t}^{2}+\sigma^{2}}$ in eq. (4) of the corresponding paper.

Specifically, in the code part, I do not understand how you implement $\nabla_{x_{t}}\log\mu(y\mid x_{t})=\frac{A^{H}(y-Ax_{t})}{\gamma_{t}^{2}+\sigma^{2}}$. I guess the code for this is lines 136-152 in main.py, which I copy as follows for reference

                    # get measurements for current estimate
                    meas = forward_operator(normalize(samples, estimated_mvue))
                    # compute gradient, i.e., gradient = A_adjoint * ( y - Ax_hat )
                    # here A_adjoint also involves the sensitivity maps, hence the pointwise multiplication
                    # also convert to real value since the ``complex'' image is a real-valued two channel image
                    meas_grad = torch.view_as_real(torch.sum(self._ifft(meas-ref) * torch.conj(maps), axis=1)).permute(0,3,1,2)
                    # re-normalize, since measuremenets are from a normalized estimate
                    meas_grad = unnormalize(meas_grad, estimated_mvue)
                    # convert to float incase it somehow became double
                    meas_grad = meas_grad.type(torch.cuda.FloatTensor)
                    meas_grad /= torch.norm( meas_grad )
                    meas_grad *= torch.norm( p_grad )
                    meas_grad *= self.config['mse']

                    # combine measurement gradient, prior gradient and noise
                    samples = samples + step_size * (p_grad - meas_grad) + noise

However, from the above code, I do not quite understand the correspondence between _meas_grad_ with the formula $\nabla_{x_{t}}\log\mu(y\mid x_{t})=\frac{A^{H}(y-Ax_{t})}{\gamma_{t}^{2}+\sigma^{2}}$. For example, it seems that only the FFT matrix is considered in A, and I do not see the values of $\sigma^2$, and in particular the value of $\gamma_{t}^2$.  Why there is operations like meas_grad /= torch.norm( meas_grad ) and meas_grad *= torch.norm( p_grad ), and something like meas_grad = unnormalize(meas_grad, estimated_mvue)? 

I find it difficult to understand the code. I tried to write the code on my own as eq (4) but it does not work. The resultant $\nabla_{x_{t}}\log\mu(y\mid x_{t})=\frac{A^{H}(y-Ax_{t})}{\gamma_{t}^{2}+\sigma^{2}}$ is really big even if I choose a large $\gamma_{t}^{2}$ at the very beginning. 
 
Could you please add some explanations on how you realize eq. (4) step by step in your code? It would be really appreciated if you can use some mathematical equations to illustrate it. 

Thank you very much.

Best
Meng

**Comment by mariusarvinte** [comment]
Hi,

Thanks for your questions and raising this important issue. As a general answer, I think the following two clarifications should help regarding Eq. (4): (i) $\sigma$ is actually assumed to be zero during inference (we do not estimate SNR) and (ii) $\gamma_t$ is not a fixed numerical constant, rather it also depends on the current residual and includes a re-normalization of the residual to match the magnitude of the score.

The code is correct and should reproduce the performance in the paper when running out-of-the-box.

More details:
- Sensitivity maps, FFT and masking are all considered in $A$. See https://github.com/utcsilab/csgm-mri-langevin/blob/main/main.py#L111 and further https://github.com/utcsilab/csgm-mri-langevin/blob/main/utils.py#L147 for the exact implementation of $A$.

- Given that $\sigma$ is assumed zero, a simpler expression for the measurement gradient is (up to a multiplicative constant, assumed to be absorbed in $\gamma_t$ :

$$
\nabla_{x_t} \log \mu (y | x_t) = A^H (y - Ax_t)
$$

- The exact expression of $\gamma_t$ is the following:

$$
\gamma_t = step size \cdot \frac{|| s(x_t; \theta) ||_2}{|| A^H (y - Ax_t)||_2}
$$

where $step size$ is given by `step_size` in the above code and is a constant value. You can recognize the above expressions from lines 146-151 in what you pasted above.

Let me know if this answers your questions.

**Comment by mengxiangming** [comment]
Thank you so much for your prompt and detailed reply! 

Now I understand it mostly, except for a few additional issues. 
> 
> $\gamma_t$ is not a fixed numerical constant, rather it also depends on the current residual and includes a re-normalization of the residual to match the magnitude of the score.

I see. So it is a heuristic choice to match the magnitude of the score, right? 


> 
> * The exact expression of $\gamma_t$ is the following:
> 
> $\gamma_t=stepsize\cdot \frac{||s(x_t;θ)||_2}{ ||A^H(y−Ax_t)||_2}$
> 
> where $step size$ is given by `step_size` in the above code and is a constant value. You can recognize the above expressions from lines 146-151 in what you pasted above.
> 

Maybe there is a typo in the above formula? From the code, I think the exact expression of $\gamma_t$ is
 
$\gamma_t=\sqrt{\frac{||A^H(y−Ax_t)||_2}{||s(x_t;θ)||_2}}$

or, equivalently,

$\gamma^2_t=\frac{||A^H(y−Ax_t)||_2}{||s(x_t;θ)||_2}$

The stepsize is multiplied outside, i.e.,
`samples = samples + step_size * (p_grad - meas_grad) + noise`


The last question I have is about the value of self.config['mse'] in the code. The value meas_grad is multiplied by self.config['mse'] as follows in line 149:
`meas_grad *= self.config['mse']`
I noted that it is set to be self.config['mse']=5.0 as default in the configuration files in all cases. 

What does self.config['mse'] stand for? Why such scaling value is needed?


Thank you very much!

Best,
Meng
